### ISSUE TITLE
Harbor v2.0 run install.sh script have a big problem -- prepare script failed after setting certs.

### ISSUE THREAD

**Issue by sober-wang** [goharbor/harbor#13461]
./install.sh meet some problem
docker version 19.03
docker-compose version 1.25.0, build 0a186604
docker-py version: 4.1.0
CPython version: 3.7.4
OpenSSL version: OpenSSL 1.1.0l  10 Sep 2019

[Step 4]: preparing harbor configs ...
prepare base dir is set to /usr/local/harbor-2.0
Clearing the configuration file: /config/log/logrotate.conf
Clearing the configuration file: /config/log/rsyslog_docker.conf
Generated configuration file: /config/log/logrotate.conf
Generated configuration file: /config/log/rsyslog_docker.conf
Traceback (most recent call last):
  File "main.py", line 15, in <module>
    cli()
  File "/usr/lib/python3.6/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/src/app/commands/prepare.py", line 47, in prepare
    prepare_nginx(config_dict)
  File "/usr/src/app/utils/nginx.py", line 23, in prepare_nginx
    render_nginx_template(config_dict)
  File "/usr/src/app/utils/nginx.py", line 56, in render_nginx_template
    prepare_nginx_certs(config_dict['cert_key_path'], config_dict['cert_path'])
  File "/usr/src/app/utils/nginx.py", line 42, in prepare_nginx_certs
    shutil.copy2(host_ngx_cert_key_path, real_key_path)
  File "/usr/lib/python3.6/shutil.py", line 257, in copy2
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "/usr/lib/python3.6/shutil.py", line 120, in copyfile
    with open(src, 'rb') as fsrc:
IsADirectoryError: [Errno 21] Is a directory: '/hostfs/data/cert'
[root@registry harbor-2.0]# 

How can find the dir /hostfs ?
How disabled the configerution?

**Comment by reasonerjt** [comment]
This error is thrown from the prepare container and `/hostfs` was mapped to the data volume.

Looks like the certs were not copied to the target directory?

Please attach the harbor.yml or docker-compose.yml after you copied the cert files.

**Comment by samuelchen** [comment]
I am still facing this issue:

Docker Desktop 2.3.0.3(45519)
MacOS: Mojave 10.14.6
Images: goharbor/prepare                     v2.2.1                                           cc1c8af914f9

I checked the `prepare` script. The mapping is `-v /:/hostfs`

```
❯ ./prepare
prepare base dir is set to /Users/samuel/work/harbor/harbor
Clearing the configuration file: /config/log/logrotate.conf
Clearing the configuration file: /config/log/rsyslog_docker.conf
Clearing the configuration file: /config/portal/nginx.conf
Generated configuration file: /config/portal/nginx.conf
Generated configuration file: /config/log/logrotate.conf
Generated configuration file: /config/log/rsyslog_docker.conf
Traceback (most recent call last):
  File "main.py", line 15, in <module>
    cli()
  File "/usr/lib/python3.6/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/src/app/commands/prepare.py", line 47, in prepare
    prepare_nginx(config_dict)
  File "/usr/src/app/utils/nginx.py", line 23, in prepare_nginx
    render_nginx_template(config_dict)
  File "/usr/src/app/utils/nginx.py", line 56, in render_nginx_template
    prepare_nginx_certs(config_dict['cert_key_path'], config_dict['cert_path'])
  File "/usr/src/app/utils/nginx.py", line 42, in prepare_nginx_certs
    shutil.copy2(host_ngx_cert_key_path, real_key_path)
  File "/usr/lib/python3.6/shutil.py", line 263, in copy2
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "/usr/lib/python3.6/shutil.py", line 120, in copyfile
    with open(src, 'rb') as fsrc:
FileNotFoundError: [Errno 2] No such file or directory: '/hostfs/Users/samuel/work/harbor/certs/reg.mbp.com.key'
❯ ls /Users/samuel/work/harbor/certs/reg.mbp.com.key
/Users/samuel/work/harbor/certs/reg.mbp.com.key
```

**Comment by sahramelissa** [comment]
Hi guys :)

If you look at prepare script you will see 

docker run ... -v /:/hostfs 

change it to docker run ... -v /pathOfYourCert/certs:/hostfs 

That will fix the problem :)

**Comment by shanwen** [comment]
I use -v /data/cert:/hostfs/data/cert fix the problem

**Comment by gao1399677** [comment]
> I use -v /data/cert:/hostfs/data/cert fix the problem

Resolved! Its puzzled me half a day

**Comment by wy65701436** [comment]
@sober-wang please try with solution provided by @sahramelissa, and reopen if you still have problem. Thanks @sahramelissa

**Comment by buglib** [comment]
> I use -v /data/cert:/hostfs/data/cert fix the problem

It works! Thanks!

**Comment by Garen-in-bush** [comment]
> Hi guys :)
> 
> If you look at prepare script you will see
> 
> docker run ... -v /:/hostfs
> 
> change it to docker run ... -v /pathOfYourCert/certs:/hostfs
> 
> # NOT WORKING Run prepare script
> docker run --rm -v $input_dir:/input -v $data_path:/data -v $harbor_prepare_path:/compose_location -v $config_dir:/config -v /:/hostfs --privileged goharbor/prepare:v2.1.5 prepare $@
> 
> # Run prepare script with the right Path
> **docker run --rm -v $input_dir:/input -v $data_path:/data -v $harbor_prepare_path:/compose_location -v $config_dir:/config -v /pathOfYourCert/certs:/hostfs --privileged goharbor/prepare:v2.1.5 prepare $@**
> 
> That will fix the problem :)

it still does not work

**Comment by ninoksmn** [comment]
Hi, 

I am setting up v2.8.0 and I am getting this error: 
# ./prepare
prepare base dir is set to /root/harbor
Traceback (most recent call last):
  File "/usr/src/app/main.py", line 15, in <module>
    cli()
  File "/usr/lib/python3.10/site-packages/click/core.py", line 1137, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3.10/site-packages/click/core.py", line 1062, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3.10/site-packages/click/core.py", line 1668, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3.10/site-packages/click/core.py", line 763, in invoke
    return __callback(*args, **kwargs)
  File "/usr/src/app/commands/prepare.py", line 35, in prepare
    config_dict = parse_yaml_config(conf, with_notary=with_notary, with_trivy=with_trivy)
  File "/usr/src/app/utils/configs.py", line 241, in parse_yaml_config
    config_dict['logger_sweeper_duration'] = js_config["logger_sweeper_duration"]
KeyError: 'logger_sweeper_duration'

Tried with adding -v /data/cert:/hostfs/data/cert
but that did not fix a problem.

**Comment by TKINGNAMe** [comment]
应该是 关于harbor的配置的错误 这里的两个路径需要指定到具体的证书的目录
There should be a configuration error with harbor where the two paths need to be specified to the directory of the specific certificate
  certificate: /data/harbor/ca/server.crt
  private_key: /data/harbor/ca/server.key

**Comment by MeinAJ** [comment]
> I use -v /data/cert:/hostfs/data/cert fix the problem

it works!

**Comment by Mitrajeet-Golsangi** [comment]
Still here in 2025 with Harbor v2.13.1 and this still seems to be a problem! I think this is some internal issue, since it is taking in the certificate path correctly but the key path does not work. I have tried all the solutions from previous or similar issues. None seem to be working. Please if anyone has a fix for this in Ubuntu 22.04 server, that would be great!
